File: src/systems/output.ts
import type { System, LLMStreamEvent } from '#src/core/types';

export const OutputSystem: System = (snapshot, event) => {
  const loop = snapshot.getComponent('agent', 'loop');
  const currentStep = loop?.step ?? 0;

  switch (event.type) {
    case 'llm.thinking_delta':
      return [{ type: 'effect.render', content: event.delta, stream: 'thinking' }];
    case 'llm.text_delta':
      return [{ type: 'effect.render', content: event.delta, stream: 'text' }];
    case 'llm.tool_call':
      return [{ type: 'effect.render', content: `[Tool: ${event.tool_call.function.name}]`, stream: 'text' }];
    case 'llm.stream_end':
      return [{ type: 'effect.emit', event: { type: 'system.step_end', step: currentStep } }];
    case 'llm.stream_error':
      return [{ type: 'effect.render', content: `Error: ${event.error}`, stream: 'text' }];
    default:
      return [];
  }
};

File: crates/alice-core/src/systems/output.rs
use crate::components::LoopComponent;
use crate::effect::{Effect, StreamType};
use crate::event::{Event, LLMStreamEvent, SystemEvent};
use crate::world::Snapshot;

pub fn output_system<C>(snapshot: &Snapshot<&C>, event: &Event) -> Vec<Effect>
where
    C: crate::world::HasComponent<LoopComponent>,
{
    match event {
        Event::LLMStream(LLMStreamEvent::ThinkingDelta { delta }) => {
            vec![Effect::Render {
                content: delta.clone(),
                stream: StreamType::Thinking,
            }]
        }
        Event::LLMStream(LLMStreamEvent::TextDelta { delta }) => {
            vec![Effect::Render {
                content: delta.clone(),
                stream: StreamType::Text,
            }]
        }
        Event::LLMStream(LLMStreamEvent::ToolCall { tool_call }) => {
            vec![Effect::Render {
                content: format!("[Tool: {}]", tool_call.function.name),
                stream: StreamType::Text,
            }]
        }
        Event::LLMStream(LLMStreamEvent::StreamEnd { .. }) => {
            let step = snapshot.get::<LoopComponent>().step;
            vec![
                Effect::Render {
                    content: "\n".into(),
                    stream: StreamType::Text,
                },
                Effect::Emit {
                    event: Event::System(SystemEvent::StepEnd { step }),
                },
            ]
        }
        Event::LLMStream(LLMStreamEvent::StreamError { error }) => {
            vec![Effect::Render {
                content: format!("Error: {error}"),
                stream: StreamType::Text,
            }]
        }
        _ => vec![],
    }
}
